library/core/src/num/f16.rs

```diff
@@ -628,12 +791,52 @@ impl f16 {
     /// ```
     #[inline]
     #[unstable(feature = "f16", issue = "116909")]
+    #[rustc_const_unstable(feature = "const_float_bits_conv", issue = "72447")]
     #[must_use = "this returns the result of the operation, without modifying the original"]
-    pub fn to_bits(self) -> u16 {
-        // SAFETY: `u16` is a plain old datatype so we can always... uh...
-        // ...look, just pretend you forgot what you just read.
-        // Stability concerns.
-        unsafe { mem::transmute(self) }
+    pub const fn to_bits(self) -> u16 {
+        // SAFETY: `u16` is a plain old datatype so we can always transmute to it.
+        // ...sorta.
+        //
+        // It turns out that at runtime, it is possible for a floating point number
+        // to be subject to a floating point mode that alters nonzero subnormal numbers
+        // to zero on reads and writes, aka "denormals are zero" and "flush to zero".
+        //
+        // And, of course evaluating to a NaN value is fairly nondeterministic.
+        // More precisely: when NaN should be returned is knowable, but which NaN?
+        // So far that's defined by a combination of LLVM and the CPU, not Rust.
+        // This function, however, allows observing the bitstring of a NaN,
+        // thus introspection on CTFE.
+        //
+        // In order to preserve, at least for the moment, const-to-runtime equivalence,
+        // we reject any of these possible situations from happening.
+        #[rustc_const_unstable(feature = "const_float_bits_conv", issue = "72447")]
+        const fn ct_f16_to_u16(ct: f16) -> u16 {
+            // FIXME(f16_f128): we should use `.classify()` like `f32` and `f64`, but we don't
+            // yet want to rely on that on all platforms (ABI issues). So just classify the
+            // bits instead.
+
+            // SAFETY: this direction is a POD transmutation. We just can't return it unless
+            // it is normal, infinite, or zero.
+            let bits = unsafe { mem::transmute::<f16, u16>(ct) };
+            match f16::classify_bits(bits) {
+                FpCategory::Nan => {
+                    panic!("const-eval error: cannot use f16::to_bits on a NaN")
+                }
+                FpCategory::Subnormal => {
+                    panic!("const-eval error: cannot use f16::to_bits on a subnormal number")
+                }
+                FpCategory::Infinite | FpCategory::Normal | FpCategory::Zero => bits,
+            }
+        }
```

Comment:
So the reason that I did it the other way around is that if we transmute a float, or do an operation on it otherwise, LLVM may change its mind about what the bits *were*, which makes me leery of classifying the bits post-facto like this.

Reply:
I do think that this would be better, but unfortunately calling `.classify()` is just nondeterministic on x86 because of ABI bugs. So while this depends on LLVM's consistency, I think it is better than producing a probably wrong answer (and the results are correct at this time on at least x86 and aarch64). At least to get things off the ground.

Anyway that's what the FIXME is supposed to be for, but happy adjust as you prefer.

Reply:
Right.

Can we be more specific than "ABI issues"? Having done a lot of research on them, I feel like they're a giant umbrella of handwavium that people shove stuff under. "`classify` is nondet on x86, even before we involve the MXCSR" is fine.

Reply:
Const `to_bits` isn't strictly needed at this time of course, I just included it because it unblocks a surprising number of tests (especially clippy), either directly or through other `const` functions that rely on it.

Reply:
Missed your comment, I updated this to be more exact.